### PR TITLE
Create project README [9297]

### PR DESCRIPTION
**Pull request by JLBuenoLopez**
Translate README and update information according to the uncrustify.cfg file

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,400 @@
+# Code Style for *eprosima Fast DDS*
+
+This document describes the code style for the [*eprosima Fast DDS* project](https://github.com/eProsima/Fast-DDS) in order to improve the code readability and maintainability.
+
+This code style is enforced through unit tests run automatically every time a pull request is done using [*uncrustify*](https://github.com/uncrustify/uncrustify).
+Latest results from these tests can be found [here](http://jenkins.eprosima.com:8080/job/fastdds_github_uncrustify/lastBuild/).
```
Comment (lauramg15) [inline README.md]:
```suggestion
The latest results from these tests can be found [here](http://jenkins.eprosima.com:8080/job/fastdds_github_uncrustify/lastBuild/).
```

File: README.md [hunk 2]
```diff
@@ -0,0 +1,400 @@
...
+
+### Horizontal Whitespace
+
+There should not be trailing whitespace at the end of the line.
+
+#### General
+
+When declaring variables, there must be only one space between type and variable name.
+If several variables are declared in the same line, there must be a space after the comma, but not before.
+This also applies in function declarations and callings.
```
Comment (lauramg15) [inline README.md]:
```suggestion
This also applies to function declarations and callings.
```

File: README.md [hunk 3]
```diff
@@ -0,0 +1,400 @@
...
+if (!error)
+    return success;
+
+// Incorrect
+if (!error) return success;
+```
+
+### Pointer and Reference Expressions
+
+When declaring a pointer or reference variable, the `*` or `&` must be always adjacent to the type and introduce a space before the variable name.
```
Comment (lauramg15) [inline README.md]:
```suggestion
When declaring a pointer or reference variable, the `*` or `&` must be always adjacent to the type and separated of the variable name by a space.
```

File: README.md [hunk 4]
```diff
@@ -0,0 +1,400 @@
...
+This also applies when declaring a prototype and defining a function or method.
+
+In case of defining double pointers (pointer to pointer), there should be no space between the asterisks `*`.
```
Comment (lauramg15) [inline README.md]:
```suggestion
In the case of defining double pointers (pointer to pointer), there should be no space between the asterisks `*`.
```

File: README.md [hunk 5]
```diff
@@ -0,0 +1,400 @@
...
+int& x;
+
+const DomainParticipant* get_participant() const;
+```
+
+### Lambda Expressions
+
+Format parameters and bodies as for any other function, and capture lists like other comma-separated lists.
+
+For by-reference captures, do not leave a space between the ampersand `&` and the variable name.
```
Comment (lauramg15) [inline README.md]:
```suggestion
In by-reference captures, do not leave a space between the ampersand `&` and the variable name.
```

File: README.md [hunk 6]
```diff
@@ -0,0 +1,400 @@
...
+char* p;
+int** ptr;
+int& x;
+
+const DomainParticipant* get_participant() const;
+```
+
+### Lambda Expressions
+
+Format parameters and bodies as for any other function, and capture lists like other comma-separated lists.
```
Comment (lauramg15) [inline README.md]:
```suggestion
The format for parameters and lambda body should be as for any other function, and the capture lists like other comma-separated lists. 
```

File: README.md [hunk 7]
```diff
@@ -0,0 +1,400 @@
...
+        // Lambda Body
+    };
+```
+
+### Class Format
+
+Any base class name should be on the same line as the subclass name, subject to the 120-column limit.
+
+The braces should open and close each in its own line.
+This also applies for structures `struct`, unions `union` and enumerates `enum`.
```
Comment (lauramg15) [inline README.md]:
```suggestion
This also applies to structures `struct`, unions `union` and enumerates `enum`.
```

File: README.md [hunk 8]
```diff
@@ -0,0 +1,400 @@
...
+            const DomainParticipantQos& qos,
+            DomainParticipantListener* listener,
+            const StatusMask& mask = StatusMask::all());
+};
+```
+
+### Constructor Initializer List
+
+Constructor initializer parameters should be each in one line.
+Except the first parameter which is preceded by a colon `:`, the rest should have the comma `,` before the name.
```
Comment (lauramg15) [inline README.md]:
```suggestion
Except for the first parameter which is preceded by a colon `:`, the rest should have the comma `,` before the name.
```

File: README.md [hunk 9]
```diff
@@ -0,0 +1,400 @@
...
+There should also be a space between the colon and the comma following the [horizontal whitespace rules](#horizontal-whitespace).
```
Comment (lauramg15) [inline README.md]:
```suggestion
There should also be a space between the colon or the comma and the parameters following the [horizontal whitespace rules](#horizontal-whitespace).
```

File: README.md [hunk 10]
```diff
@@ -0,0 +1,400 @@
...
+```
+
+### Namespace Formatting
+
+When declaring a namespace, the open brace should be in the same line as the declaration, separated one space from the namespace name.
+This is the only exception in the code style where the opening brace is not in its own line.
+
+Namespaces do not add an extra level of indentation.
+
+There should be a comment after the closing brace of the namespace, even more if the scope of the namespace is great enough and there may be doubts about which namespace is being closed.
```
Comment (lauramg15) [inline README.md]:
```suggestion
There should be a comment after the closing brace of the namespace, as if the scope of the namespace is great enough there may be doubts about which namespace is being closed.
```

File: README.md [hunk 11]
```diff
@@ -0,0 +1,400 @@
...
+
+} // ip
+} // network
+```
+
+### Templates
+
+[//]: # (The uncrustify configuration file does not enforce that the template parameters should be each in a different line.)
+
+Template declaration should be in a different line from the entity it modifies.
```
Comment (lauramg15) [inline README.md]:
```suggestion
The template declaration should be in a different line from the entity it modifies.
```

File: README.md [hunk 12]
```diff
@@ -0,0 +1,400 @@
...
+template<typename K, typename V>
+struct map_size_helper : public detail::pool_size_helper<detail::map_node_size<K, V>::value>
+{
+};
+```
+
+### Preprocessor Directives
+
+The hash mark `#` that starts a preprocessor directive should always be at the beginning of the line.
+Even when preprocessor directives are within the body of indented code, the directives should start at the beginning of the line.
```
Comment (lauramg15) [inline README.md]:
```suggestion
Even when preprocessor directives are within a block of indented code, the directives should start at the beginning of the line.
```

File: README.md [hunk 13]
```diff
@@ -0,0 +1,400 @@
...
+
+There should be a comment after `#endif` indicating the `#if` which is ending.
+This is even more important when the scope of the `#if` is great enough and there may be doubts.
```
Comment (lauramg15) [inline README.md]:
```suggestion
This is even more important when the scope of the `#if` is great enough, and there may be doubts.
```

File: README.md [hunk 14]
```diff
@@ -0,0 +1,400 @@
...
+    ProtocolVersion_t():
+#if HAVE_SECURITY
+        // As imposed by DDSSEC11-93
+        ProtocolVersion_t(2,3)
+#else
+        ProtocolVersion_t(2,2)
+#endif // HAVE_SECURITY
+```
+
+There must be spaces around preprocessor `##` concatenation operator.
```
Comment (lauramg15) [inline README.md]:
```suggestion
There must be spaces around the preprocessor `##` concatenation operator.
```

File: README.md [hunk 15]
```diff
@@ -0,0 +1,400 @@
...
+```
+
+## Struct vs Classes
+
+Use a struct only for passive objects that carry data; everything else is a class.
+It is advisable that structs do not provide any method, although public access methods could be accepted.
+However, as the struct members should be public, there would be no need to provide those methods.
+
+If more functionality is required, a class is more appropriate.
+If in doubt, make it a class.
```
Comment (lauramg15) [inline README.md]:
```suggestion
If there is any doubt, make it a class.
```

File: README.md [hunk 16]
```diff
@@ -0,0 +1,400 @@
...
+int s_size;
+```
+
+### Filenames
+
+File extensions `.cpp` and `.hpp` are used for source files with C++ code.
+On the other hand, use `.c` and `.h` extensions if the file is in C code.
+
+Filenames should be all lowercase and can include underscores `_`.
+However, class files which define or declare a class, should take the class name.
```
Comment (lauramg15) [inline README.md]:
```suggestion
However, class files that define or declare a class should take the class name.
```

File: README.md [hunk 17]
```diff
@@ -0,0 +1,400 @@
...
+
+### Type Names
+
+Type names start with a capital letter and have a capital letter for each new word, with no underscores.
+This convention is followed by all types: classes, structs, type aliases, enums and type template parameters.
```
Comment (lauramg15) [inline README.md]:
```suggestion
This convention is followed by all types: classes, structs, type aliases, enums, and type template parameters.
```

File: README.md [hunk 18]
```diff
@@ -0,0 +1,400 @@
...
+Usually this convention is called *CamelCase*.
```
Comment (lauramg15) [inline README.md]:
```suggestion
Usually, this convention is called *CamelCase*.
```

File: README.md [hunk 19]
```diff
@@ -0,0 +1,400 @@
...
+
+### Variable Names
+
+The names of variables (including function parameters) and data members are all lowercase, with underscores between words.
+Data members of classes (but not structs) additionally have trailing underscore.
```
Comment (lauramg15) [inline README.md]:
```suggestion
Data members of classes (but not structs) additionally have a trailing underscore.
```

File: README.md [hunk 20]
```diff
@@ -0,0 +1,400 @@
...
+
+All data members should be private.
+Accesors and mutators (get and set functions) will take the name of the variable they affect without the trailing underscore.
```
Comment (lauramg15) [inline README.md]:
```suggestion
Accessors and mutators (get and set functions) will take the name of the variable they affect without the trailing underscore.
```

**Review by lauramg15 — CHANGES_REQUESTED** [review]

**Comment by MiguelCompany** [comment]
I think the whole @eProsima/team should review this

File: README.md [hunk 21]
```diff
@@ -0,0 +1,400 @@
...
+
+## Language and Standard
+
+*eprosima Fast DDS* project is mostly written in C++.
+The Standard followed is C++14.
+
+## Formatting
+
+The code style is formalized in the [uncrustify.cfg](uncrustify.cfg) file for the uncrustify tool.
+Below, the more important rules that are enforced are listed.
```
Comment (IkerLuengo) [inline README.md]:
```suggestion
The most important rules that are enforced are listed below.
```

File: README.md [hunk 22]
```diff
@@ -0,0 +1,400 @@
...
+
+This code style is enforced through unit tests run automatically every time a pull request is done using [*uncrustify*](https://github.com/uncrustify/uncrustify).
+Latest results from these tests can be found [here](http://jenkins.eprosima.com:8080/job/fastdds_github_uncrustify/lastBuild/).
+These tests only run over those files where changes have been made.
+Over time, this code style will apply to all source files.
+
+## Language and Standard
+
+*eprosima Fast DDS* project is mostly written in C++.
+The Standard followed is C++14.
```
Comment (IkerLuengo) [inline README.md]:
```suggestion
The followed Standard is C++14.
```

File: README.md [hunk 23]
```diff
@@ -0,0 +1,400 @@
...
+
+Non-ASCII characters should be rare and must use UTF-8 formatting.
+
+### Indentation
+
+Spaces are used for indentation instead of tabs.
+
+Block indentation increases +4 spaces each time a new block or block-like construct is opened.
+When the block ends, the indent returns to the previous indent level.
+If line-wrapping occurs an extra indentation of +8 spaces should be added.
```
Comment (IkerLuengo) [inline README.md]:
```suggestion
If line-wrapping occurs, an extra indentation of +8 spaces should be added.
```

File: README.md [hunk 24]
```diff
@@ -0,0 +1,400 @@
...
+
+### Vertical Whitespace
+
+There must be one blank line between function or method declarations.
+
+There must also be one blank line before and after access specifiers (`public`, `private` and `protected`) within a class definition.
+
+Within a method or function declaration, there should be blank lines separating different functionalities.
+It should be convenient to start each functionality block with a comment.
```
Comment (IkerLuengo) [inline README.md]:
```suggestion
It is convenient to start each functionality block with a comment.
```

File: README.md [hunk 25]
```diff
@@ -0,0 +1,400 @@
...
+
+### Horizontal Whitespace
+
+There should not be trailing whitespace at the end of the line.
```
Comment (IkerLuengo) [inline README.md]:
```suggestion
There should not be any trailing whitespace at the end of the line.
```

Alternatives:
```
There should be no trailing whitespace at the end of the line.
```
```
There should not be trailing whitespaces at the end of the line.
```

File: README.md [hunk 26]
```diff
@@ -0,0 +1,400 @@
...
+
+#### General
+
+When declaring variables, there must be only one space between type and variable name.
```
Comment (IkerLuengo) [inline README.md]:
```suggestion
When declaring variables, there must be only one space between the type and the variable name.
```

File: README.md [hunk 27]
```diff
@@ -0,0 +1,400 @@
...
+```
+
+#### Line-end Comments
+
+Lines that are non-obvious should get a comment at the end of the line.
+These end-of-line comments should be separated from the code by a space.
+
+### Function Declarations and Definitions
+
+To improve parameter readability, each parameter must be in its own line, even the first one and even if there is only one parameter.
```
Comment (IkerLuengo) [inline README.md]:
```suggestion
To improve parameter readability, each parameter must be in its own line, even the first one, and even if there is only one parameter.
```

File: README.md [hunk 28]
```diff
@@ -0,0 +1,400 @@
...
+The line-wrapping indentation is to be used in this case.
+Never indent to the opening parenthesis level, the first parameter, etc.
```
Comment (IkerLuengo) [inline README.md]: The first parameter, if the rules are followed, should be on its own line and indented with 8 spaces, just as the remaining parameters. This wording may cause some confusion. I'd simplify it for the sake of clarity.
```suggestion
Never indent to the opening parenthesis level.
```

File: README.md (hunk 3 above)
Comment (IkerLuengo) [reply README.md]: May I suggest `separated from` instead?

File: README.md [hunk 29]
```diff
@@ -0,0 +1,400 @@
...
+
+// Incorrect
+if (!error)
+    return success;
+
+// Incorrect
+if (!error) return success;
+```
+
+### Pointer and Reference Expressions
```
Comment (IkerLuengo) [inline README.md:201]: There is an exception when defining several pointers or references in the same line. In this case the `*` or `&` must be adjacent to the name:
```
int *value, *key;
```

File: README.md [hunk 30]
```diff
@@ -0,0 +1,400 @@
...
+
+### Lambda Expressions
+
+Format parameters and bodies as for any other function, and capture lists like other comma-separated lists.
+
+For by-reference captures, do not leave a space between the ampersand `&` and the variable name.
+
+There should not be a space separating the closing square bracket `]` from the opening parenthesis `(` when an argument list is present in the lambda expression.
+
+Lamba bodies should be indented following the indentation rules and both the opening and close braces should be each alone in their own line.
```
Comment (IkerLuengo) [inline README.md]: For coherence:
```suggestion
Lamba bodies should be indented following the indentation rules and both the opening and closing braces should be each alone in their own line.
```

File: README.md (hunk 10 above)
Comment (IkerLuengo) [reply README.md]:
```suggestion
There should be a comment after the closing brace of the namespace, as if the scope of the namespace is great enough, there may be doubts about which namespace is being closed.
```

File: README.md [hunk 31]
```diff
@@ -0,0 +1,400 @@
...
+
+```c++
+#define PASTE2( x, y) x ## y
+```
+
+## Struct vs Classes
+
+Use a struct only for passive objects that carry data; everything else is a class.
+It is advisable that structs do not provide any method, although public access methods could be accepted.
+However, as the struct members should be public, there would be no need to provide those methods.
```
Comment (IkerLuengo) [inline README.md]: I'd be more strict and request all data in the struct to be public. The only case I can think of for non-public data is if the struct needs to keep an internal state, and in that case, it should be a class.
I don't know if uncrustify can enforce this, but we still should be able to state it in the style guide.

File: README.md [hunk 32]
```diff
@@ -0,0 +1,400 @@
...
+            long count) { row_count_ = count; }
+
+    long row_count() const { return row_count_; }
+
+    long& row_count() { return row_count_; }
+
+private:
+
+    long row_count_;
+};
```
Comment (IkerLuengo) [inline README.md]:
```suggestion
class Table
{
public:

    void row_count(
            long count)
    {
        row_count_ = count;
    }

    long row_count() const
    {
        return row_count_;
    }

    long& row_count()
    {
        return row_count_;
    }

private:

    long row_count_;
};
```

**Review by IkerLuengo — CHANGES_REQUESTED** [review]

File: README.md (hunk 29 above)
Comment (JLBuenoLopez) [reply README.md:201]: Introduced

File: README.md (hunk 3 above)
Comment (JLBuenoLopez) [reply README.md]: I have opted for @IkerLuengo solution

File: README.md (hunk 29 above)
Comment (JLBuenoLopez-eProsima) [reply README.md:183]: Introduced

File: README.md (hunk 31 above)
Comment (JLBuenoLopez) [reply README.md]: Done.

**Review by IkerLuengo — DISMISSED** [review]
:+1:

File: README.md [hunk 33]
```diff
@@ -0,0 +1,412 @@
...
+}
+```
+
+#### Templates
+
+There should be a space between closing angles `>`.
+
+```c++
+std::map<std::string, std::vector<DataWriterImpl*> > writers_;
+```
```
Comment (LuisGP) [inline README.md]: I think this is the ugliest thing anybody can write. Eases confusion with the `>` operator, and doesn't clarify anything.

File: README.md (hunk 29 above)
Comment (LuisGP) [reply README.md:201]: I disagree with @IkerLuengo as the `exception` breaks the rule, no exceptions should be allowed. Indeed, there should be forbidden to declare several variables in the same line. Always.

File: README.md [hunk 34]
```diff
@@ -0,0 +1,412 @@
...
+This also applies when declaring a prototype and defining a function or method.
+
+In the case of defining double pointers (pointer to pointer), there should be no space between the asterisks `*`.
+
+```c++
+char* p;
+int** ptr;
+int& x;
+
+int *value, *key;
```
Comment (LuisGP) [inline README.md]: No exceptions!

File: README.md [hunk 35]
```diff
@@ -0,0 +1,412 @@
...
+Except for the first instance, these keywords should be preceded by a blank line.
+There should be a blank line after these keywords.
+
+```c++
+class DomainParticipant : public Entity
+{
+public: // No indentation
+                                                // Blank line
+    RTPS_DllAPI ReturnCode_t get_qos(
+        DomainParticipantQos& qos) const;
```
Comment (LuisGP) [inline README.md]:
```suggestion
            DomainParticipantQos& qos) const;
```

File: README.md [hunk 36]
```diff
@@ -0,0 +1,412 @@
...
+        ProtocolVersion_t(2,3)
+#else
+        ProtocolVersion_t(2,2)
+#endif // HAVE_SECURITY
+```
+
+There must be spaces around the preprocessor `##` concatenation operator.
+
+```c++
+#define PASTE2( x, y) x ## y
```
Comment (LuisGP) [inline README.md]:
```suggestion
#define PASTE2(x, y) x ## y
```

**Review by LuisGP — CHANGES_REQUESTED** [review]

File: README.md (hunk 34 above)
Comment (MiguelCompany) [reply README.md]: I agree with @LuisGP here. I think we should add a rule saying that we don't allow several declarations on the same line.

File: README.md (hunk 33 above)
Comment (MiguelCompany) [reply README.md]: Writing both angle brackets together (i.e. without a space in the middle), could also cause confusion with the `>>` operator, and is illegal on some old compilers. That said, support for `A<B<param>>` (without space) was added to the standard for C++11.

I see that we don't have a policy regarding closing parenthesis, so I think we should decide on both.

Should it be `my_function(my_adapter(param))` or `my_function(my_adapter(param) )` ?

My vote would go towards not using the space, being consistent with `my_function(param)`

Comment (LuisGP) [reply README.md]: We force at least C++11 compliance, so the lack of the space between the angle brackets should not be a problem.
Hence, my vote goes to not using the space as well.

File: README.md (hunk 36 above)
Comment (JLBuenoLopez) [reply README.md]: Done

File: README.md (hunk 35 above)
Comment (JLBuenoLopez) [reply README.md]: Thanks for finding it out!

File: README.md (hunk 34 above)
Comment (JLBuenoLopez) [reply README.md]: The rule has been included.

File: README.md (hunk 33 above)
Comment (JLBuenoLopez) [reply README.md]: This rule has been modified. It also includes the rule about nested parentheses. uncrustify.cfg file updated accordingly.

**Review by LuisGP — DISMISSED** [review]

**Review by lauramg15 — DISMISSED** [review]

**Review by MiguelCompany — APPROVED** [review]

**Pull request merged**